### ISSUE TITLE
[bug]: final steps of windows install fails with `PermissionError: [WinError 5] Access is denied`

### ISSUE THREAD

**Issue by diablodale** [invoke-ai/InvokeAI#2811]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### OS

Windows

### GPU

cuda

### VRAM

8GB

### Setup

* Windows 10 and Windows 11
* fresh install of python v3.9.13 64-bit, pip upgraded to v23.0.1
* InvokeAI-installer-v2.3.1.post1
* NVIDIA RTX2070 Super 8gb
* intel i7-10875h w/ 32gb

### Repro

1. download installer zip and unzip
2. doubleclick on `WinLongPathsEnabled.reg` and agree
3. reboot
4. run `install.bat`
5. choose install options, models, and allow for the GB's of downloads and installs
6. Wait for error near the final steps of install

### Result

Cascading python exceptions. Here is the first exception (full log output is in later section)

```
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\shutil.py:627 in _rmtree_unsafe         │
│                                                                                                  │
│    624 │   │   │   _rmtree_unsafe(fullname, onerror)                                             │
│    625 │   │   else:                                                                             │
│    626 │   │   │   try:                                                                          │
│ ❱  627 │   │   │   │   os.unlink(fullname)                                                       │
│    628 │   │   │   except OSError:                                                               │
│    629 │   │   │   │   onerror(os.unlink, fullname, sys.exc_info())                              │
│    630 │   try:                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
PermissionError: [WinError 5] Access is denied: 'C:\\Users\\dale\\AppData\\Local\\Temp\\invokeai-installer-tmpjof__zyv\\Lib\\site-packages\\charset_normalizer\\md.cp39-win_amd64.pyd'
```

Also note that if I dir that file...it is there.
```
C:\Users\dale>dir C:\Users\dale\AppData\Local\Temp\invokeai-installer-tmpjof__zyv\Lib\site-packages\charset_normalizer\md.cp39-win_amd64.pyd
 Volume in drive C has no label.
 Volume Serial Number is F48B-4C47

 Directory of C:\Users\dale\AppData\Local\Temp\invokeai-installer-tmpjof__zyv\Lib\site-packages\charset_normalizer

02/25/2023  05:14 PM            10,752 md.cp39-win_amd64.pyd
               1 File(s)         10,752 bytes
               0 Dir(s)  434,245,230,592 bytes free
```

### Expected

Clean install. No errors.

### Additional context

Is installer/python trying to remove the `C:\Users\dale\AppData\Local\Temp\invokeai-installer-tmpjof__zyv\...` folder(s) yet it is still using the pdy's within it?

```
** DOWNLOADING SUPPORT MODELS **
Installing bert tokenizer...
Installing SD1 clip model...
Installing SD2 clip model...
Installing models from RealESRGAN...
Installing RealESRGAN model file https://github.com/xinntao/Real-ESRGAN/releases/download/v0.2.5.0/realesr-general-x4v3.pth......exists
Installing RealESRGANwdn model file https://github.com/xinntao/Real-ESRGAN/releases/download/v0.2.5.0/realesr-general-wdn-x4v3.pth......exists
Installing GFPGAN models...
Installing GFPGAN weights model file https://github.com/TencentARC/GFPGAN/releases/download/v1.3.0/GFPGANv1.4.pth......exists
Installing GFPGAN weights model file https://github.com/xinntao/facexlib/releases/download/v0.1.0/detection_Resnet50_Final.pth......exists
Installing GFPGAN weights model file https://github.com/xinntao/facexlib/releases/download/v0.2.2/parsing_parsenet.pth......exists
Installing CodeFormer model file...
Installing CodeFormer model file https://github.com/sczhou/CodeFormer/releases/download/v0.1.0/codeformer.pth......exists
Installing clipseg model for text-based masking...
Installing model for NSFW content detection...
AutoFeatureExtractor...
StableDiffusionSafetyChecker...
Installing stabilityai VAE...
* vae-ft-mse-840000-ema-pruned.ckpt: complete file found. Skipping.

** DOWNLOADING DIFFUSION WEIGHTS **

** INVOKEAI INSTALLATION SUCCESSFUL **
If you installed manually from source or with 'pip install': activate the virtual environment
then run one of the following commands to start InvokeAI.

Web UI:
   invokeai --web # (connect to http://localhost:9090)
   invokeai --web --host 0.0.0.0 # (connect to http://your-lan-ip:9090 from another computer on the local network)

Command-line interface:
   invokeai

If you installed using an installation script, run:
  C:\Users\dale\invokeai/invoke.bat

Add the '--help' argument to see all of the command-line switches available for use.

╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\shutil.py:627 in _rmtree_unsafe         │
│                                                                                                  │
│    624 │   │   │   _rmtree_unsafe(fullname, onerror)                                             │
│    625 │   │   else:                                                                             │
│    626 │   │   │   try:                                                                          │
│ ❱  627 │   │   │   │   os.unlink(fullname)                                                       │
│    628 │   │   │   except OSError:                                                               │
│    629 │   │   │   │   onerror(os.unlink, fullname, sys.exc_info())                              │
│    630 │   try:                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
PermissionError: [WinError 5] Access is denied: 'C:\\Users\\dale\\AppData\\Local\\Temp\\invokeai-installer-tmpjof__zyv\\Lib\\site-packages\\charset_normalizer\\md.cp39-win_amd64.pyd'

During handling of the above exception, another exception occurred:

╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\tempfile.py:805 in onerror              │
│                                                                                                  │
│   802 │   │   │   │   │   resetperms(path)                                                       │
│   803 │   │   │   │   │                                                                          │
│   804 │   │   │   │   │   try:                                                                   │
│ ❱ 805 │   │   │   │   │   │   _os.unlink(path)                                                   │
│   806 │   │   │   │   │   # PermissionError is raised on FreeBSD for directories                 │
│   807 │   │   │   │   │   except (IsADirectoryError, PermissionError):                           │
│   808 │   │   │   │   │   │   cls._rmtree(path)                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
PermissionError: [WinError 5] Access is denied: 'C:\\Users\\dale\\AppData\\Local\\Temp\\invokeai-installer-tmpjof__zyv\\Lib\\site-packages\\charset_normalizer\\md.cp39-win_amd64.pyd'

During handling of the above exception, another exception occurred:

╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\weakref.py:667 in _exitfunc             │
│                                                                                                  │
│   664 │   │   │   │   │   │   # threads) the following is the only line in                       │
│   665 │   │   │   │   │   │   # this function which might trigger creation                       │
│   666 │   │   │   │   │   │   # of a new finalizer                                               │
│ ❱ 667 │   │   │   │   │   │   f()                                                                │
│   668 │   │   │   │   │   except Exception:                                                      │
│   669 │   │   │   │   │   │   sys.excepthook(*sys.exc_info())                                    │
│   670 │   │   │   │   │   assert f not in cls._registry                                          │
│                                                                                                  │
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\weakref.py:591 in __call__              │
│                                                                                                  │
│   588 │   │   otherwise return None"""                                                           │
│   589 │   │   info = self._registry.pop(self, None)                                              │
│   590 │   │   if info and not self._shutdown:                                                    │
│ ❱ 591 │   │   │   return info.func(*info.args, **(info.kwargs or {}))                            │
│   592 │                                                                                          │
│   593 │   def detach(self):                                                                      │
│   594 │   │   """If alive then mark as dead and return (obj, func, args, kwargs);                │
│                                                                                                  │
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\tempfile.py:820 in _cleanup             │
│                                                                                                  │
│   817 │                                                                                          │
│   818 │   @classmethod                                                                           │
│   819 │   def _cleanup(cls, name, warn_message):                                                 │
│ ❱ 820 │   │   cls._rmtree(name)                                                                  │
│   821 │   │   _warnings.warn(warn_message, ResourceWarning)                                      │
│   822 │                                                                                          │
│   823 │   def __repr__(self):                                                                    │
│                                                                                                  │
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\tempfile.py:816 in _rmtree              │
│                                                                                                  │
│   813 │   │   │   else:                                                                          │
│   814 │   │   │   │   raise                                                                      │
│   815 │   │                                                                                      │
│ ❱ 816 │   │   _shutil.rmtree(name, onerror=onerror)                                              │
│   817 │                                                                                          │
│   818 │   @classmethod                                                                           │
│   819 │   def _cleanup(cls, name, warn_message):                                                 │
│                                                                                                  │
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\shutil.py:759 in rmtree                 │
│                                                                                                  │
│    756 │   │   │   onerror(os.path.islink, path, sys.exc_info())                                 │
│    757 │   │   │   # can't continue even if onerror hook returns                                 │
│    758 │   │   │   return                                                                        │
│ ❱  759 │   │   return _rmtree_unsafe(path, onerror)                                              │
│    760                                                                                           │
│    761 # Allow introspection of whether or not the hardening against symlink                     │
│    762 # attacks is supported on the current platform                                            │
│                                                                                                  │
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\shutil.py:624 in _rmtree_unsafe         │
│                                                                                                  │
│    621 │   │   │   except OSError:                                                               │
│    622 │   │   │   │   onerror(os.path.islink, fullname, sys.exc_info())                         │
│    623 │   │   │   │   continue                                                                  │
│ ❱  624 │   │   │   _rmtree_unsafe(fullname, onerror)                                             │
│    625 │   │   else:                                                                             │
│    626 │   │   │   try:                                                                          │
│    627 │   │   │   │   os.unlink(fullname)                                                       │
│                                                                                                  │
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\shutil.py:624 in _rmtree_unsafe         │
│                                                                                                  │
│    621 │   │   │   except OSError:                                                               │
│    622 │   │   │   │   onerror(os.path.islink, fullname, sys.exc_info())                         │
│    623 │   │   │   │   continue                                                                  │
│ ❱  624 │   │   │   _rmtree_unsafe(fullname, onerror)                                             │
│    625 │   │   else:                                                                             │
│    626 │   │   │   try:                                                                          │
│    627 │   │   │   │   os.unlink(fullname)                                                       │
│                                                                                                  │
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\shutil.py:624 in _rmtree_unsafe         │
│                                                                                                  │
│    621 │   │   │   except OSError:                                                               │
│    622 │   │   │   │   onerror(os.path.islink, fullname, sys.exc_info())                         │
│    623 │   │   │   │   continue                                                                  │
│ ❱  624 │   │   │   _rmtree_unsafe(fullname, onerror)                                             │
│    625 │   │   else:                                                                             │
│    626 │   │   │   try:                                                                          │
│    627 │   │   │   │   os.unlink(fullname)                                                       │
│                                                                                                  │
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\shutil.py:629 in _rmtree_unsafe         │
│                                                                                                  │
│    626 │   │   │   try:                                                                          │
│    627 │   │   │   │   os.unlink(fullname)                                                       │
│    628 │   │   │   except OSError:                                                               │
│ ❱  629 │   │   │   │   onerror(os.unlink, fullname, sys.exc_info())                              │
│    630 │   try:                                                                                  │
│    631 │   │   os.rmdir(path)                                                                    │
│    632 │   except OSError:                                                                       │
│                                                                                                  │
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\tempfile.py:808 in onerror              │
│                                                                                                  │
│   805 │   │   │   │   │   │   _os.unlink(path)                                                   │
│   806 │   │   │   │   │   # PermissionError is raised on FreeBSD for directories                 │
│   807 │   │   │   │   │   except (IsADirectoryError, PermissionError):                           │
│ ❱ 808 │   │   │   │   │   │   cls._rmtree(path)                                                  │
│   809 │   │   │   │   except FileNotFoundError:                                                  │
│   810 │   │   │   │   │   pass                                                                   │
│   811 │   │   │   elif issubclass(exc_info[0], FileNotFoundError):                               │
│                                                                                                  │
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\tempfile.py:816 in _rmtree              │
│                                                                                                  │
│   813 │   │   │   else:                                                                          │
│   814 │   │   │   │   raise                                                                      │
│   815 │   │                                                                                      │
│ ❱ 816 │   │   _shutil.rmtree(name, onerror=onerror)                                              │
│   817 │                                                                                          │
│   818 │   @classmethod                                                                           │
│   819 │   def _cleanup(cls, name, warn_message):                                                 │
│                                                                                                  │
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\shutil.py:759 in rmtree                 │
│                                                                                                  │
│    756 │   │   │   onerror(os.path.islink, path, sys.exc_info())                                 │
│    757 │   │   │   # can't continue even if onerror hook returns                                 │
│    758 │   │   │   return                                                                        │
│ ❱  759 │   │   return _rmtree_unsafe(path, onerror)                                              │
│    760                                                                                           │
│    761 # Allow introspection of whether or not the hardening against symlink                     │
│    762 # attacks is supported on the current platform                                            │
│                                                                                                  │
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\shutil.py:610 in _rmtree_unsafe         │
│                                                                                                  │
│    607 │   │   with os.scandir(path) as scandir_it:                                              │
│    608 │   │   │   entries = list(scandir_it)                                                    │
│    609 │   except OSError:                                                                       │
│ ❱  610 │   │   onerror(os.scandir, path, sys.exc_info())                                         │
│    611 │   │   entries = []                                                                      │
│    612 │   for entry in entries:                                                                 │
│    613 │   │   fullname = entry.path                                                             │
│                                                                                                  │
│ C:\Users\dale\AppData\Local\Programs\Python\Python39\lib\shutil.py:607 in _rmtree_unsafe         │
│                                                                                                  │
│    604 # version vulnerable to race conditions                                                   │
│    605 def _rmtree_unsafe(path, onerror):                                                        │
│    606 │   try:                                                                                  │
│ ❱  607 │   │   with os.scandir(path) as scandir_it:                                              │
│    608 │   │   │   entries = list(scandir_it)                                                    │
│    609 │   except OSError:                                                                       │
│    610 │   │   onerror(os.scandir, path, sys.exc_info())                                         │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
NotADirectoryError: [WinError 267] The directory name is invalid: 'C:\\Users\\dale\\AppData\\Local\\Temp\\invokeai-installer-tmpjof__zyv\\Lib\\site-packages\\charset_normalizer\\md.cp39-win_amd64.pyd'
Press any key to continue . . .
```

**Comment by lstein** [comment]
Despite the error message, I believe that InvokeAI installed correctly on your system. The crash is occurring at a cleanup step that happens after a successful install.

Could you go to the `invokeai` directory located at `C:\Users\dale\invokeai` and launch the `invoke` .bat file?

@ebr The error is happening when TemporaryDirectory tries to remove the virtual environment it created during the bootstrap phase. I don't see anything wrong in the code that should cause this.

**Comment by ebr** [comment]
InvokeAI did indeed install correctly.

The `PermissionError` is caused by a race condition in Python 3.9 on Windows, preventing cleanup of temporary directory [1]. Python 3.10 adds the `ignore_cleanup_errors` argument [2] to the `tempfile.TemporaryDirectory()`. But for earlier versions we unfortunately do not have the option of squelching it [3].
We are already recommending Python3.10 for any new installs; I don't know if there's much else we can do to fix this aside from documentation/FAQ...

@diablodale do you have a specific need to use Python 3.9, or could you install 3.10 instead?

[1] https://bugs.python.org/issue42796
[2] https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryDirectory
[3] https://github.com/invoke-ai/InvokeAI/pull/2448/commits/9e22ed5c12eedce819e58100ac4e015fce90ac09

**Comment by diablodale** [comment]
Hi. Yes, I have successfully used `invoke.bat` after getting the OP exceptions. I have not noticed errant behavior of the webui or inference that I could associated with the exceptions.

I use Python 3.9.x for my diverse set of projects, e.g.

* compiling/linking to python headers/libs for 3rd party C++ code/SDKs that do not yet support python 3.10
* minimizing differences from AWS Lambda; they do not support 3.10

When the race can't be fixed with client code, then adding some text to https://invoke-ai.github.io/InvokeAI/installation/010_INSTALL_AUTOMATED/#other-problems for this issue work for me. Or now that this issue exists, someone can also query for it and find it here.

**Comment by ebr** [comment]
likely fixed by [#3123](https://github.com/invoke-ai/InvokeAI/pull/3123)

**Comment by T3chRomanc3r** [comment]
Basically tried to install Invoke on my second SSD drive by choosing custom directory, but got OSError [WinError 5].
Tried launching .bat with Administrator rights, but same issue persists.

Full log of installation:
```
InvokeAI will be installed in C:\Users\User\invokeai
Would you like to pick a different location? [y/n] (n): y

Please select the destination directory for the installation: [E:\InvokeAI-Installer]:

E:\InvokeAI-Installer\Invoke

InvokeAI will be installed in E:\InvokeAI-Installer\Invoke
Would you like to pick a different location? [y/n] (n): n

Requirement already satisfied: pip in e:\invokeai-installer\invoke\.venv\lib\site-packages (21.2.4)
Collecting pip
  Using cached pip-23.2.1-py3-none-any.whl (2.1 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 21.2.4
    Uninstalling pip-21.2.4:
ERROR: Could not install packages due to an OSError: [WinError 5] Access is denied: 'e:\\invokeai-installer\\invoke\\.venv\\scripts\\pip.exe'
Check the permissions.

Traceback (most recent call last):
  File "E:\InvokeAI-Installer\lib\main.py", line 50, in <module>
    inst.install(**args.__dict__)
  File "E:\InvokeAI-Installer\lib\installer.py", line 174, in install
    self.venv = self.app_venv()
  File "E:\InvokeAI-Installer\lib\installer.py", line 147, in app_venv
    pip["install", "--upgrade", "pip"] & FG
  File "C:\Users\User\AppData\Local\Temp\invokeai-installer-tmpjw5tc7dm\Lib\site-packages\plumbum\commands\modifiers.py", line 145, in __rand__
    cmd(
  File "C:\Users\User\AppData\Local\Temp\invokeai-installer-tmpjw5tc7dm\Lib\site-packages\plumbum\commands\base.py", line 113, in __call__
    return self.run(args, **kwargs)[1]
  File "C:\Users\User\AppData\Local\Temp\invokeai-installer-tmpjw5tc7dm\Lib\site-packages\plumbum\commands\base.py", line 252, in run
    return p.run()
  File "C:\Users\User\AppData\Local\Temp\invokeai-installer-tmpjw5tc7dm\Lib\site-packages\plumbum\commands\base.py", line 215, in runner
    return run_proc(p, retcode, timeout)
  File "C:\Users\User\AppData\Local\Temp\invokeai-installer-tmpjw5tc7dm\Lib\site-packages\plumbum\commands\processes.py", line 299, in run_proc
    return _check_process(proc, retcode, timeout, stdout, stderr)
  File "C:\Users\User\AppData\Local\Temp\invokeai-installer-tmpjw5tc7dm\Lib\site-packages\plumbum\commands\processes.py", line 17, in _check_process
    proc.verify(retcode, timeout, stdout, stderr)
  File "C:\Users\User\AppData\Local\Temp\invokeai-installer-tmpjw5tc7dm\Lib\site-packages\plumbum\machines\base.py", line 27, in verify
    raise ProcessExecutionError(
plumbum.commands.processes.ProcessExecutionError: Unexpected exit code: 1
Command line: | 'E:\InvokeAI-Installer\Invoke\.venv\Scripts\pip.exe' install --upgrade pip
Press any key to continue . . .
```
